packages/js/components/src/experimental-select-control/select-control.tsx

```diff
@@ -113,42 +128,84 @@ function SelectControl< ItemType = DefaultItemType >( {
 		getItemProps,
 		selectItem,
 		selectedItem: singleSelectedItem,
+		openMenu,
+		closeMenu,
 	} = useCombobox< ItemType | null >( {
 		inputValue,
 		items: filteredItems,
+		selectedItem: null,
 		itemToString: getItemLabel,
 		// eslint-disable-next-line @typescript-eslint/ban-ts-comment
 		// @ts-ignore
-		onStateChange: ( { inputValue: value, type, selectedItem } ) => {
+		onStateChange: ( changes ) => {
```

Comment:
Is it possible to move all (or at least most of) the logic in the `onStateChange` to the `stateReducer`? I think this would simply this component a good bit if we could just look to one area for changes.

Reply:
Unfortunately I don't think so, I purposely left all the function calls that calls our state handlers in the `onStateChange`, as it felled out of place to trigger other changes within the `stateReducer` function, I think that function is mostly meant to stay pure, without any side effects. Pretty well all of the items left in the `onStateChange` are side effects.

Reply:
I think as is, the state reducer is fairly broken.  Here are a couple of the issues I've run into:

In a multiple select control, hit enter on an item.  The input value gets updated to the selected item's text along with adding the item:
<img width="197" alt="Screen Shot 2022-09-29 at 2 44 47 PM" src="https://user-images.githubusercontent.com/10561050/193147924-2672b57a-8a8c-44cc-a32e-54f13c0dec2c.png">

In a single select control, select a value.  Afterwards remove a character or more and blur the input.  See that the input value gets removed despite still having a selected value.
<img width="332" alt="Screen Shot 2022-09-29 at 2 44 15 PM" src="https://user-images.githubusercontent.com/10561050/193147929-23af9715-b404-4afc-a6af-b5cf65b4ac3b.png">
<img width="296" alt="Screen Shot 2022-09-29 at 2 44 21 PM" src="https://user-images.githubusercontent.com/10561050/193147928-c6436a24-b4c4-4843-b2ed-c916341cd2d1.png">

Reply:
I would think that we could leave the `stateReducer` very simple and keep our side effects as they were in the `onStateChange`.  For example:

```js
		stateReducer: ( state, actionAndChanges ) => {
			const { changes } = actionAndChanges;
			const newChanges = stateReducer
				? stateReducer( state, actionAndChanges )
				: {};
			return {
				...changes,
				...newChanges,
			};
```

It seems like there's a lot of manipulation around the input value in the `stateReducer` in your current changes.  If your use case needs to be able to control the input value, maybe we should rely on Downshift's internal `inputValue` instead of handling our own.  That way, we can just have a `stateReducer` and won't need duplicate logic in the `onStateChange`.

>as it felled out of place to trigger other changes within the stateReducer function, I think that function is mostly meant to stay pure, without any side effects

I understand the reasoning here and this would be my preference as well, though I'm not sure it would cause any real problems if we need to manipulate our own state within this reducer.